saleor/plugins/vatlayer/__init__.py

```diff
@@ -89,3 +98,131 @@ def get_taxed_shipping_price(shipping_price, taxes):
+    """
+    voucher = checkout_info.voucher
+    if not voucher or voucher.type == VoucherType.SHIPPING:
+        return line_price
+
+    line_quantity = checkout_line_info.line.quantity
+    total_discount_amount = checkout_info.checkout.discount_amount
+    line_total_price = line_price * line_quantity
+    currency = checkout_info.checkout.currency
+
+    # if the checkout has only one line, the hole discount amount will be applied
+    # to this line
+    if len(lines) == 1:
+        return (
+            line_total_price - Money(total_discount_amount, currency)
+        ) / line_quantity
+
+    # if the checkout has more lines we need to propagate the discount amount
+    # proportionally to total prices of items
+    lines_total_prices = [
+        base_calculations.calculate_base_line_unit_price(
+            line_info,
+            checkout_info.channel,
+            discounts,
+        ).price_with_discounts.amount
+        * line_info.line.quantity
+        for line_info in lines
+        if line_info.line.id != checkout_line_info.line.id
+    ]
+
+    total_price = sum(lines_total_prices) + line_total_price.amount
+
+    last_element = lines[-1].line.id == checkout_line_info.line.id
+    if last_element:
+        discount_amount = _calculate_discount_for_last_element(
+            lines_total_prices, total_price, total_discount_amount, currency
+        )
+    else:
+        discount_amount = quantize_price(
+            line_total_price.amount / total_price * total_discount_amount, currency
+        )
+    return quantize_price(
+        (line_total_price - Money(discount_amount, currency)) / line_quantity, currency
+    )
+
+
+def apply_order_discount_to_order_unit_price(
+    order: "Order",
+    order_line: "OrderLine",
+):
+    """Calculate the order line price with discounts.
+
+    Include the entire order voucher discount.
+    """
+    if (
+        not order.voucher_id
+        or order.voucher.type == VoucherType.SHIPPING  # type: ignore
+    ):
+        return order_line.base_unit_price
+
+    currency = order.currency
+    total_discount_amount = get_total_order_discount_excluding_shipping(order).amount
+    line_total_price = order_line.base_unit_price * order_line.quantity
+
+    # if the order has only one line, the hole discount amount will be applied
+    # to this line
+    if order.lines.count() == 1:
+        return (
+            line_total_price - Money(total_discount_amount, currency)
+        ) / order_line.quantity
+
+    order_lines = order.lines.all()
+
+    # if the order has more lines we need to propagate the discount amount
+    # proportionally to total prices of items
+    lines_total_prices = [
+        line.base_unit_price.amount * line.quantity
+        for line in order_lines
+        if line.id != order_line.id
+    ]
+    total_price = sum(lines_total_prices) + line_total_price.amount
+
+    last_element = order_lines.last().pk == order_line.pk  # type: ignore
+    if last_element:
```

Comment:
This tells me we're implementing this logic on the wrong layer. Logic executed for a single line should not have to consult any of the other lines, let alone fetch them all from the database. We need a function that operates on the entire order, that can loop over the lines and distribute whatever discount there is.